src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,484 @@
+    builder_->BeginDataflowBlock();
+    // accept bindings in the original block
+    for (const auto& binding : block->bindings) {
+      this->VisitBinding(binding);
+    }
+
+    // generate backward bindings and the return value
+    return_expr_ = BackwardBindingGenerator().Generate(this->builder_, GetRef<DataflowBlock>(block),
+                                                       this->require_grads_, this->target_var_);
+
+    return builder_->EndBlock();
+  }
+
+ private:
+  // check that the target should be a VarNode, not DataflowVarNode
+  // and a scalar of type "DynTensorType"
+  static void CheckTarget(const Expr& e) {
+    CHECK(e->IsInstance<VarNode>()) << "The differentiation target must be a Var";
+    CHECK(!e->IsInstance<DataflowVarNode>()) << "The differentiation target is not an output node";
+    auto sinfo = GetStructInfoAs<TensorStructInfoNode>(e);
+    CHECK(sinfo != nullptr) << "The differentiation target must be a Tensor";
+    CHECK(sinfo->shape.defined() && sinfo->shape.as<ShapeExprNode>())
+        << "Error when getting the shape of the differentiation target";
+    CHECK(sinfo->shape.as<ShapeExprNode>()->values.size() == 0)
+        << "The differentiation target must be scalar";
+  }
+
+  // Check the type of the input var should be Tensor of floating point dtype, or Tuple of that
+  static void CheckFloatTensorType(const StructInfo& sinfo, const String& name_hint) {
+    if (auto* tuple_sinfo = sinfo.as<TupleStructInfoNode>()) {
+      for (auto item : tuple_sinfo->fields) {
+        CheckFloatTensorType(item, name_hint);
+      }
+    } else if (auto* tensor_sinfo = sinfo.as<TensorStructInfoNode>()) {
+      CHECK(tensor_sinfo->dtype.is_float() || tensor_sinfo->dtype.is_float16() ||
+            tensor_sinfo->dtype.is_bfloat16())
+          << "Only Tensors of floating point dtype can require gradients, but the dtype of Var "
+          << name_hint << " is " << tensor_sinfo->dtype;
+    } else {
+      LOG(FATAL) << "The input Var " << name_hint << " is neither a Tensor nor a Tuple of Tensors";
+    }
+  }
+
+  // Checks every Var in require_grads:
+  // 1. there should be no duplicate var
+  // 2. every var should be a parameter of the function
+  // 3. the type of the input var should be Tensor of floating point dtype, or Tuple of that
+  static void CheckRequireGrads(const Array<Var>& require_grads, const Array<Var>& func_params,
+                                const String& func_name) {
+    std::unordered_set<Var, ObjectPtrHash, ObjectPtrEqual> var_set;
+    for (const auto& var : require_grads) {
+      CHECK(std::find(func_params.begin(), func_params.end(), var) != func_params.end())
+          << "There is no Var named " << var->name_hint() << " in the parameters of the function "
+          << func_name;
+      CHECK(var_set.count(var) == 0) << "Var " << var->name_hint() << " appears more than once";
+      var_set.emplace(var);
+
+      CheckFloatTensorType(GetStructInfo(var), var->name_hint());
```

Comment:
Consider making it a util function like

https://github.com/mlc-ai/relax/blob/relax/src/relax/transform/utils.h#L41

Reply:
Okay, let me create another PR in mlc-ai/relax